lib/cadet/assessments/assessments.ex

```diff
@@ -1413,30 +1414,75 @@ defmodule Cadet.Assessments do
     end
   end
 
-  @spec update_submission_status_and_xp_bonus(Submission.t()) ::
+  @spec update_submission_status(Submission.t()) ::
           {:ok, Submission.t()} | {:error, Ecto.Changeset.t()}
-  defp update_submission_status_and_xp_bonus(submission = %Submission{}) do
-    assessment = submission.assessment
-    assessment_conifg = Repo.get_by(AssessmentConfig, id: assessment.config_id)
+  defp update_submission_status(submission = %Submission{}) do
+    submission
+    |> Submission.changeset(%{status: :submitted, submitted_at: Timex.now()})
+    |> Repo.update()
+  end
 
-    max_bonus_xp = assessment_conifg.early_submission_xp
-    early_hours = assessment_conifg.hours_before_early_xp_decay
+  @spec update_xp_bonus(Submission.t()) ::
+          {:ok, Submission.t()} | {:error, Ecto.Changeset.t()}
+  defp update_xp_bonus(submission = %Submission{id: submission_id}) do
+    # to ensure backwards compatibility
+    if submission.xp_bonus == 0 do
+      assessment = submission.assessment
+      assessment_conifg = Repo.get_by(AssessmentConfig, id: assessment.config_id)
 
-    xp_bonus =
-      if Timex.before?(Timex.now(), Timex.shift(assessment.open_at, hours: early_hours)) do
-        max_bonus_xp
-      else
-        # This logic interpolates from max bonus at early hour to 0 bonus at close time
-        decaying_hours = Timex.diff(assessment.close_at, assessment.open_at, :hours) - early_hours
-        remaining_hours = Enum.max([0, Timex.diff(assessment.close_at, Timex.now(), :hours)])
-        proportion = if(decaying_hours > 0, do: remaining_hours / decaying_hours, else: 1)
-        bonus_xp = round(max_bonus_xp * proportion)
-        Enum.max([0, bonus_xp])
-      end
+      max_bonus_xp = assessment_conifg.early_submission_xp
+      early_hours = assessment_conifg.hours_before_early_xp_decay
 
-    submission
-    |> Submission.changeset(%{status: :submitted, xp_bonus: xp_bonus})
-    |> Repo.update()
+      ans_xp =
+        Answer
+        |> where(submission_id: ^submission_id)
+        |> order_by(:question_id)
+        |> group_by([a], a.id)
+        |> select([a], %{
+          # grouping by submission, so s.xp_bonus will be the same, but we need an
+          # aggregate function
+          total_xp: sum(a.xp) + sum(a.xp_adjustment)
```

Comment:
rethink about what you are querying before editing the code. As mentioned, there is only one answer and there is not need to sum.